src/main/scala/com/fulcrumgenomics/util/Sequences.scala

```diff
@@ -150,15 +152,74 @@ object Sequences {
   }
 
   /** Reverse complements a string of bases. */
-  def revcomp(s: String): String = SequenceUtil.reverseComplement(s)
+  def revcomp(s: String): String = {
+    val bs = s.getBytes
+    revcomp(bs)
+    new String(bs)
+  }
 
   /** Returns the sequence that is the complement of the provided sequence. */
   def complement(s: String): String = {
     val bs = s.getBytes
-    forloop(from=0, until=bs.length) { i =>  bs(i) = SequenceUtil.complement(bs(i)) }
+    complement(bs)
     new String(bs)
   }
 
+  /** Reverse complements an array of bases in place. See [[complement()]] for how bases are complemented. */
+  def revcomp(bs: Array[Byte]): Unit = {
+    SequenceUtil.reverse(bs, 0, bs.length)
+    complement(bs)
+  }
+
+  /**
+    * Complements the bases in the array in place. `U` and `u` are always complemented to `A` and `a` respectively.
+    * IUPAC ambiguity codes are handled appropriately.  Any byte that is not a valid upper- or lower-case DNA base,
+    * RNA base or IUPAC code will cause an exception to be thrown.
+    *
+    * @param bs
+    */
+  def complement(bs: Array[Byte]): Unit = {
+    forloop (from=0, until=bs.length) { i =>
+      val complement: Byte = (bs(i): @switch) match {
+        case 'A' => 'T'
+        case 'C' => 'G'
+        case 'G' => 'C'
+        case 'T' => 'A'
+        case 'U' => 'A'
+        case 'M' => 'K'
+        case 'R' => 'Y'
+        case 'W' => 'S'
+        case 'S' => 'W'
+        case 'Y' => 'R'
+        case 'K' => 'M'
+        case 'V' => 'B'
+        case 'H' => 'D'
+        case 'D' => 'H'
+        case 'B' => 'V'
+        case 'N' => 'N'
+        case 'a' => 't'
```

Comment:
you could just check if the character is lower case before the match, then upper case it for the match, then lower case the result if necessary.

Reply:
Since it's less likely to have a lot of IUPAC bases, if you stick with listing all the upper and lower case bases, how about moving the a/c/g/t/u cases to under the A/C/G/T/U cases?

Reply:
Heh, my first implementation did the upper/lower casing, but I think it's more efficient to have a single match/switch.  Also the order shouldn't really matter as it's getting compiled to a switch which is just a big jump table.